runtime-common/stdlib/math/bcmath.cpp

```diff
@@ -0,0 +1,923 @@
+    }
+  }
+
+  if (lsign < 0 && rsign > 0) {
+    if (bc_comp_impl(lhs, lint, ldot, lfrac, lscale, rhs, rint, rdot, rfrac, rscale, 1000000000) <= 0) {
+      return bc_sub_positive(rhs, rint, rdot, rfrac, rscale, lhs, lint, ldot, lfrac, lscale, scale, 1);
+    } else {
+      return bc_sub_positive(lhs, lint, ldot, lfrac, lscale, rhs, rint, rdot, rfrac, rscale, scale, -1);
+    }
+  }
+
+  if (lsign < 0 && rsign < 0) {
+    return bc_add_positive(lhs, lint, ldot, lfrac, lscale, rhs, rint, rdot, rfrac, rscale, scale, -1);
+  }
+
+  php_assert(0);
+}
+
+} // namespace legacy
+
+int bc_comp(const bcmath_impl_::BcNum &lhs, const bcmath_impl_::BcNum &rhs, int scale) noexcept {
+  return legacy::bc_comp_impl(lhs.str.c_str(), lhs.n_int, lhs.n_dot, lhs.n_frac, lhs.n_scale, rhs.str.c_str(), rhs.n_int, rhs.n_dot, rhs.n_frac, rhs.n_scale,
+                              scale);
+}
+
+string bc_mul_positive(const bcmath_impl_::BcNum &lhs, const bcmath_impl_::BcNum &rhs, int scale, int sign) noexcept {
+  return legacy::bc_mul_positive_impl(lhs.str.c_str(), lhs.n_int, lhs.n_dot, lhs.n_frac, lhs.n_scale, rhs.str.c_str(), rhs.n_int, rhs.n_dot, rhs.n_frac,
+                                      rhs.n_scale, scale, sign);
+}
+
+string bc_div_positive(const bcmath_impl_::BcNum &lhs, const bcmath_impl_::BcNum &rhs, int scale, int sign) noexcept {
+  return legacy::bc_div_positive_impl(lhs.str.c_str(), lhs.n_int, lhs.n_dot, lhs.n_frac, lhs.n_scale, rhs.str.c_str(), rhs.n_int, rhs.n_dot, rhs.n_frac,
+                                      rhs.n_scale, scale, sign);
+}
+
+string bc_add(const bcmath_impl_::BcNum &lhs, const bcmath_impl_::BcNum &rhs, int scale) noexcept {
+  return legacy::bc_add_impl(lhs.str.c_str(), lhs.n_sign, lhs.n_int, lhs.n_dot, lhs.n_frac, lhs.n_scale, rhs.str.c_str(), rhs.n_sign, rhs.n_int, rhs.n_dot,
+                             rhs.n_frac, rhs.n_scale, scale);
+}
+
+string bc_sub(const bcmath_impl_::BcNum &lhs, const bcmath_impl_::BcNum &rhs, int scale) noexcept {
+  return legacy::bc_add_impl(lhs.str.c_str(), lhs.n_sign, lhs.n_int, lhs.n_dot, lhs.n_frac, lhs.n_scale, rhs.str.c_str(), (-1) * rhs.n_sign, rhs.n_int,
+                             rhs.n_dot, rhs.n_frac, rhs.n_scale, scale);
+}
+
+} // namespace
+
+std::pair<bcmath_impl_::BcNum, bool> bcmath_impl_::bc_parse_number(const string &num) noexcept {
+  BcNum bc_num;
+  bc_num.str = num;
+  bool success = legacy::bc_parse_number_impl(num, bc_num.n_sign, bc_num.n_int, bc_num.n_dot, bc_num.n_frac, bc_num.n_scale) >= 0;
+  return {bc_num, success};
+}
+
+void f$bcscale(int64_t scale) noexcept {
+  if (scale < 0) {
+    MathLibContext::get().bc_scale = 0;
+  } else {
+    MathLibContext::get().bc_scale = static_cast<int>(scale);
+  }
+}
+
+string f$bcdiv(const string &lhs_str, const string &rhs_str, int64_t scale) noexcept {
+  if (scale == std::numeric_limits<int64_t>::min()) {
+    scale = MathLibContext::get().bc_scale;
```

Comment:
You can store a reference to math lib context in a local variable to prevent calling it multiple times in a single function.